packages/extension-url-mapping/src/setup-app.ts

```diff
@@ -0,0 +1,187 @@
+    increment: () => void
+    decrement: () => void
+}
+
+// This is safe to delete if your extension does not use state. If you aren't using this, ensure you remove the
+// `withApplicationExtensionStore` usage below as well.
+const sliceInitializer: SliceInitializer<StoreSlice> = (set) => ({
+    count: 0,
+    increment: () => set((state) => ({count: state.count + 1})),
+    decrement: () => set((state) => ({count: state.count - 1}))
+})
+
+class UrlMapping extends ApplicationExtension<Config> {
+    static readonly id = extensionMeta.id
+
+    /**
+     * Use this method to wrap or enhance your PWA-Kit application using [React higher-order components](https://legacy.reactjs.org/docs/higher-order-components.html).
+     * You can use this to add visual treatments to your application, change the props that are supplied to the application component
+     * or add things like providers and contexts to be used throughout your app.
+     */
+    extendApp<T extends React.ComponentType<T>>(
+        App: React.ComponentType<T>
+    ): React.ComponentType<T> {
+        const HOCs = [
+            // Example higher-order component, this can be safely removed.
+            sampleHOC,
+            // Optionally include state for this extension using `withApplicationExtensionStore`
+            (component: React.ComponentType<any>) =>
+                withApplicationExtensionStore(component, {
+                    id: extensionMeta.id,
+                    initializer: sliceInitializer
+                })
+        ]
+
+        return applyHOCs(App, HOCs)
+    }
+
+    /**
+     * TODO: update this comment
+     * This method is used to make changes to the PWA-Kit application routes. If your extension adds a new page to the application
+     * then you can add it to the router here. The routes passed to this method is an accrued list of routes that have been added
+     * from extensions applied before it. It is called during the `getRoutes` phase on both the server and the client.
+     *
+     * NOTE: If you instead want to modify a list of all the routes, refer to the `beforeRouteMatch` below.
+     */
+    async getRoutes({locals}: GetRoutesParams): Promise<RouteProps[]> {
+        if (!locals.originalUrl) {
+            return Promise.resolve([])
+        }
+
+        // Example of API call to get url mapping
+        const config = this.getConfig()
+        const shopperSeo = await getShopperSeoClient(locals, config)
+        const urlMapping = await shopperSeo.getUrlMapping({
+            parameters: {urlSegment: locals.originalUrl}
+        })
+        console.log('--- url mapping', urlMapping)
+
+        const requestURL = new URL(locals.originalUrl, getAppOrigin(locals))
+
+        // Returns a partial route
+        return Promise.resolve([
+            {
+                path: requestURL.pathname,
+                componentName: 'Foo',
+                exact: true
+                // component: SamplePage
+            }
+        ])
```

Comment:
As we've discussed, we were thinking that `getRoutes` can return an incomplete route like this.